library/alloc/src/vec.rs

```diff
@@ -1311,35 +1310,7 @@ impl<T> Vec<T> {
         // the hole, and the vector length is restored to the new length.
         //
         let len = self.len();
-        let start = match range.start_bound() {
-            Included(&n) => n,
-            Excluded(&n) => n + 1,
-            Unbounded => 0,
-        };
-        let end = match range.end_bound() {
-            Included(&n) => n + 1,
-            Excluded(&n) => n,
-            Unbounded => len,
-        };
-
-        #[cold]
-        #[inline(never)]
-        fn start_assert_failed(start: usize, end: usize) -> ! {
-            panic!("start drain index (is {}) should be <= end drain index (is {})", start, end);
-        }
-
-        #[cold]
-        #[inline(never)]
-        fn end_assert_failed(end: usize, len: usize) -> ! {
-            panic!("end drain index (is {}) should be <= len (is {})", end, len);
-        }
-
-        if start > end {
-            start_assert_failed(start, end);
-        }
-        if end > len {
-            end_assert_failed(end, len);
-        }
+        let Range { start, end } = self.check_range(range);
```

Comment:
It looks like this method was pretty finely tuned, do we have any existing benchmarks we can check to see whether there's any impact?

Reply:
I don't think `Vec::drain` has any benches. These are the benches for `Vec`:
https://github.com/rust-lang/rust/blob/master/library/alloc/benches/vec.rs

However, I wouldn't expect a regression. `check_range` calls functions with the [same attributes](https://github.com/rust-lang/rust/pull/75207/files/d04e6b8de5fe6bbf203c534c35e6f55e8960ab46#diff-e5a4dbd27d5eadb7a3cf746c396f5420R3092) for this reason. That's part of why I think it's better to package it with the standard library.

Reply:
If there aren't any benches then I'd consider it a bit of a speculative implementation anyways, so am happy for this to be made simpler :+1: